targets/TARGET_NORDIC/TARGET_NRF5x/trng_api.c

```diff
@@ -0,0 +1,119 @@
+ *
+ *   2. Redistributions in binary form, except as embedded into a Nordic Semiconductor ASA
+ *      integrated circuit in a product or a software update for such product, must reproduce
+ *      the above copyright notice, this list of conditions and the following disclaimer in
+ *      the documentation and/or other materials provided with the distribution.
+ *
+ *   3. Neither the name of Nordic Semiconductor ASA nor the names of its contributors may be
+ *      used to endorse or promote products derived from this software without specific prior
+ *      written permission.
+ *
+ *   4. This software, with or without modification, must only be used with a
+ *      Nordic Semiconductor ASA integrated circuit.
+ *
+ *   5. Any software provided in binary or object form under this license must not be reverse
+ *      engineered, decompiled, modified and/or disassembled.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#if defined(DEVICE_TRNG)
+
+#include "hal/trng_api.h"
+#include "hal/critical_section_api.h"
+
+#include "nrf_drv_rng.h"
+
+/* Keep track of instantiated FlashIAP objects. */
+static int nordic_trng_counter = 0;
+
+void trng_init(trng_t *obj)
+{
+    (void) obj;
+
+    /* Increment global counter inside critical section. */
+    hal_critical_section_enter();
+    nordic_trng_counter++;
+    int counter = nordic_trng_counter;
+    hal_critical_section_exit();
+
+    /* Initialize TRNG on first object only. */
+    if (counter == 1) {
+        nrf_drv_rng_init(NULL);
+    }
+}
+
+void trng_free(trng_t *obj)
+{
+    (void) obj;
+
+    /* Decrement global counter inside critical section. */
+    hal_critical_section_enter();
+    nordic_trng_counter--;
+    int counter = nordic_trng_counter;
+    hal_critical_section_exit();
+
+    /* Deinitialize TRNG when all objects have been freed. */
+    if (counter == 0) {
+        nrf_drv_rng_uninit();
+    }
+}
+
+/* Get random data from NRF5x TRNG peripheral.
+ *
+ * This implementation returns num of random bytes in range <1, length>.
+ * For parameters description see trng_api.h file.
+ */
+int trng_get_bytes(trng_t *obj, uint8_t *output, size_t length, size_t *output_length)
+{
+    (void) obj;
```

Comment:
```trng_init``` and ```trng_free``` are synchronised but not this function. Please document if there are reasons to believe this is thread safe.

Reply:
Mbed HAL functions are not supposed to be thread safe - the caller is responsible for handling thread safety.

I'm not sure why `trng_init` and `trng_free` are safe - they should actually be non-safe to be consistent with the rest of the HAL.

Reply:
@marcuschangarm Can the original author perhaps shed some light on this?

Reply:
@marcuschangarm These functions are called from ```feature/mbedtls``` and not directly by the user. Mbed TLS calls these functions as abstract API and it does not know if a particular platform requires synchronisation. Hence, it is fine to provide some synchronisation here if needed.

Reply:
Shouldn't the thread safety be handled in Mbed TLS since that is the user facing API?

Reply:
I believe yes, there is a wrapper for tls to mbed where sync can be provided. HAL does not provide at all in any case.

Reply:
@0xc0170 should I remove the synchronization from `trng_init` and `trng_free` then?

Reply:
Just talked to @sbutcher-arm and he suggested that Mbed TLS in mbed-os only supports single threaded access for now. Hence, I would suggest keeping these functions consistent with other platform implementation.

On how the synchronisation should happen in this case (once Mbed TLS starts supporting multi-threaded apps) requires study.

Reply:
Ah, ok! Thank you for looking into this.

I'll add two mutexes: one for `trng_init` and `trng_free` and one for `trng_get_bytes`.

Reply:
We shouldn't be using ```trng_get_bytes``` before ```trng_init``` is called and after ```trng_free``` is called. Hence, there should just be one mutex for these three.

Reply:
Thanks @mazimkhan for the info

Reply:
Now I'm a bit confused! 😄 

Do you want me to:

1. take out all thread safety that is currently there.
1. insert thread safety everywhere.

I'm fine with either, but the current 2/3 thread safety looks weird.

Reply:
I also raised the point of 2/3 safety earlier. My opinion is to be consistent with other platforms that don't do any thread safety here. Hence, pt.1.
We will revisit thread safety later for all platforms and may introduce something similar.

Reply:
Sounds good! I'll remove it completely then!